File: tests/cross_driver/java/src/main/java/com/secantus/smokes/TailableSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.CursorType;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.CreateCollectionOptions;
import org.bson.Document;

import java.util.concurrent.TimeUnit;

/**
 * Cross-driver tailable cursor smoke — Java.
 * <p>
 * Open a tailable cursor on a capped collection, drain the seeded
 * doc, then insert another and verify the cursor surfaces it without
 * being reopened.
 */
public final class TailableSmoke {

    private TailableSmoke() {}

    public static void main(String[] args) throws Exception {
        String uri = System.getenv("MONGODB_URI");
        if (uri == null) {
            System.err.println("MONGODB_URI not set");
            System.exit(2);
        }

        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .applyToClusterSettings(b -> b.serverSelectionTimeout(5, TimeUnit.SECONDS))
            .build();

        try (MongoClient client = MongoClients.create(settings)) {
            MongoDatabase db = client.getDatabase("tailable_xd");
            db.drop();
            db.createCollection("logs", new CreateCollectionOptions()
                .capped(true).sizeInBytes(64 * 1024));
            MongoCollection<Document> coll = db.getCollection("logs");
            coll.insertOne(new Document("_id", 1));

            try (MongoCursor<Document> cursor = coll.find(new Document())
                    .cursorType(CursorType.Tailable)
                    .cursor()) {
                Document first = cursor.next();
                if (first == null || first.getInteger("_id") != 1) {
                    System.err.printf("first: %s, want {_id: 1}%n",
                        first == null ? null : first.toJson());
                    System.exit(1);
                }

                coll.insertOne(new Document("_id", 2));

                long deadline = System.currentTimeMillis() + 5000;
                while (System.currentTimeMillis() < deadline) {
                    Document ev = cursor.tryNext();
                    if (ev != null) {
                        if (ev.getInteger("_id") != 2) {
                            System.err.printf("second: %s, want {_id: 2}%n", ev.toJson());
                            System.exit(1);
                        }
                        System.out.println("OK");
                        return;
                    }
                    Thread.sleep(100);
                }
                System.err.println("tailable cursor did not surface the new doc within 5s");
                System.exit(1);
            }
        }
    }
}

File: tests/cross_driver/node/tailable_smoke.js
// Cross-driver tailable cursor smoke — Node.

const { MongoClient } = require('mongodb');

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

async function main() {
  const uri = process.env.MONGODB_URI;
  if (!uri) fail('MONGODB_URI not set');

  const client = new MongoClient(uri, { serverSelectionTimeoutMS: 5000 });
  await client.connect();
  try {
    const db = client.db('tailable_xd');
    await db.dropDatabase().catch(() => {});
    await db.createCollection('logs', { capped: true, size: 64 * 1024 });
    const coll = db.collection('logs');
    await coll.insertOne({ _id: 1 });

    const cursor = coll.find({}, { tailable: true });
    try {
      // Drain the seeded doc.
      const first = await cursor.next();
      if (!first || first._id !== 1) fail(`first: ${JSON.stringify(first)}, want {_id: 1}`);

      // Insert another doc; tailable cursor should surface it.
      await coll.insertOne({ _id: 2 });

      const deadline = Date.now() + 5000;
      while (Date.now() < deadline) {
        const ev = await cursor.tryNext();
        if (ev) {
          if (ev._id !== 2) fail(`second: ${JSON.stringify(ev)}, want {_id: 2}`);
          console.log('OK');
          return;
        }
        await sleep(100);
      }
      fail('tailable cursor did not surface the new doc within 5s');
    } finally {
      await cursor.close();
    }
  } finally {
    await client.close();
  }
}

main().catch((err) => fail(`uncaught: ${err && err.stack ? err.stack : err}`));
